src/PowerShellEditorServices/Language/RunspaceSynchronizer.cs

```diff
@@ -0,0 +1,272 @@
+                handlerDelegate: HandleRunspaceStateChange,
+                supportEvent: true,
+                forwardEvent: false);
+
+            Activate();
+            // Trigger events
+            HandleRunspaceStateChange(sender: null, args: null);
+        }
+
+        /// <summary>
+        /// Does a thing
+        /// </summary>
+        public static void Activate()
+        {
+            SourceActionEnabled = true;
+        }
+
+        #endregion
+
+        #region Private Methods
+
+        private static void HandleRunspaceStateChange(object sender, PSEventArgs args)
+        {
+            if (!SourceActionEnabled)
+            {
+                return;
+            }
+
+            SourceActionEnabled = false;
+
+            var newOrChangedModules = new List<PSModuleInfo>();
+            List<PSModuleInfo> modules = ReflectionUtils.GetModules(sourceRunspace);
+            foreach (PSModuleInfo module in modules)
+            {
+                if (moduleCache.Add(module))
+                {
+                    newOrChangedModules.Add(module);
+                }
+            }
+
+
+            var newOrChangedVars = new List<PSVariable>();
+
+            var variables = sourceEngineIntrinsics.GetVariables();
+            foreach (var variable in variables)
+            {
+                // If the variable is a magic variable or it's type has not changed, then skip it.
+                if(POWERSHELL_MAGIC_VARIABLES.Contains(variable.Name) ||
+                    (variableCache.TryGetValue(variable.Name, out Type value) && value == variable.Value?.GetType()))
+                {
+                    continue;
+                }
+
+                // Add the variable to the cache and mark it as a newOrChanged variable.
+                variableCache[variable.Name] = variable.Value?.GetType();
+                newOrChangedVars.Add(variable);
+            }
+
+            if (newOrChangedModules.Count > 0)
+            {
+                // Import the modules in the targetRunspace with -Force
+                using (PowerShell pwsh = PowerShell.Create())
+                {
+                    pwsh.Runspace = targetRunspace;
+
+                    foreach (PSModuleInfo moduleInfo in newOrChangedModules)
+                    {
+                        if(moduleInfo.Path != null)
+                        {
+                            pwsh.AddCommand("Import-Module")
+                                .AddParameter("Name", moduleInfo.Path)
+                                .AddParameter("Force")
+                                .AddStatement();
+                        }
+                    }
+
+                    pwsh.Invoke();
+                }
+            }
+
+            if (newOrChangedVars.Count > 0)
+            {
+                // Set or update the variables.
+                foreach (PSVariable variable in newOrChangedVars)
+                {
+                    targetEngineIntrinsics.SetVariable(variable);
```

Comment:
If the debugger is currently stopped in a different session state (like a break point in a module function), this will throw all of the module's state into the global state for the completion runspace.  Also stepping through something like `Pester` will dump *a lot* of vars into completion for the rest of the session.

Reply:
Good point. Need to try out this in debugging scenario.

Reply:
Rob and I have brainstormed about this. For starters, we will skip synchronizing while the initial runspace is debugging. The belief here is that your main focus in debugging is reading, not writing. Unfortunately, this means stuff like `$TestName` and `$TestDrive` ala `Pester` will not be available in completions _while debugging_... but we think it's not critical enough for this PR.

Eventually, it would be nice to get to a point where _after_ you finish debugging, it triggers a "resync -force" which blows away the target runspace and remakes it with a clean sync so we're back to what we would expect after debugging.

Reply:
> The belief here is that your main focus in debugging is reading, not writing.

Writing at a breakpoint, along with `$Host.EnterNestedPrompt()` in general (which will also have the same problem) are the only ways to get intellisense based on a module's session state.  I use this extensively.

Reply:
I've disabled syncing while debugging.

Reply:
Sorry for the sadness, like I said, eventually I'd like to do the "trigger resync force" - both for debugging and nested prompts but it's not the highest priority for the majority of customers and I 'm not quite sure how I would approach it just yet.

Any ideas?